### PR TITLE
Remove runningFlows cache in ExecutorManager. Added unit tests for Ex…

### PR DESCRIPTION
**Pull request by jamiesjc**
Tested the changes in solo server and local mysql database. Fixed some bugs in existing test cases for ExecutorManger and JdbcExecutorLoader.

TODO: 
1. Remove queuedFlows cache and refactor the unit tests for ExecutorManager. Need to remove QueueProcessorThread as well. 
2. Refactor executorManager, inject ExecutorApiClient using Guice. This will help mock http calls to executors in unit tests. 
3. When a running flow is canceled, it takes some time for the job/flow log to show up. Need to fix this issue. 
4. Refactor JdbcExecutorLoaderTest. Consider using in-memory database.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java [hunk 1]
```diff
@@ -219,6 +219,27 @@ public ExecutableFlow fetchExecutableFlow(int id)
...
+      else {
+        return flows.get(0);
+      }
+    } catch (SQLException e) {
+      throw new ExecutorManagerException("Error fetching active flows by exec id", e);
+    }
+  }
+
+  @Override
   public int fetchNumExecutableFlows() throws ExecutorManagerException {
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: Looks like fetchNumExecutableFlows never bean used. Shall we remove it?

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java [hunk 2]
```diff
@@ -1410,6 +1431,68 @@ public String handle(ResultSet rs) throws SQLException {
...
   }
 
+  private static class FetchActiveExecutableFlowByExecId implements
+      ResultSetHandler<List<Pair<ExecutionReference, ExecutableFlow>>> {
+    private static String FETCH_ACTIVE_EXECUTABLE_FLOW_BY_EXECID =
+        "SELECT ex.exec_id exec_id, ex.enc_type enc_type, ex.flow_data flow_data, et.host host, "
+            + "et.port port, ax.update_time axUpdateTime, et.id executorId, et.active executorStatus"
+            + " FROM execution_flows ex"
+            + " INNER JOIN "
+            + " active_executing_flows ax ON ex.exec_id = ax.exec_id"
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: Just curious: do you know which server updates table `active_executing_flows `, web server or exeuctor, or both?

File: azkaban-common/src/main/java/azkaban/executor/ExecutorLoader.java [hunk 3]
```diff
@@ -35,6 +35,9 @@ ExecutableFlow fetchExecutableFlow(int execId)
   Map<Integer, Pair<ExecutionReference, ExecutableFlow>> fetchActiveFlows()
       throws ExecutorManagerException;
 
+  Pair<ExecutionReference, ExecutableFlow> fetchActiveFlowByExecId(int execId)
```
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorLoader.java]: No, we don't have it. When a flow is inactive, we just remove it from DB. We don't keep inactive flows info in DB.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 4]
```diff
@@ -420,23 +417,23 @@ public Executor fetchExecutor(int executorId) throws ExecutorManagerException {
...
     for (Executor executor : activeExecutors) {
       ports.add(executor.getHost() + ":" + executor.getPort());
     }
-    // include executor which were initially active and still has flows running
-    for (Pair<ExecutionReference, ExecutableFlow> running : runningFlows
-      .values()) {
-      ExecutionReference ref = running.getFirst();
-      ports.add(ref.getHost() + ":" + ref.getPort());
+    try {
+      // include executor which were initially active and still has flows running
```
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Yes.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 5]
```diff
@@ -469,8 +466,12 @@ private void loadQueuedFlows() throws ExecutorManagerException {
     if (runningCandidate != null) {
       executionIds.addAll(getRunningFlowsHelper(projectId, flowId, Lists.newArrayList(runningCandidate)));
     }
-    executionIds.addAll(getRunningFlowsHelper(projectId, flowId,
-      runningFlows.values()));
+    try {
```
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Will this confuse with getExecutableFlows()? In the future, queuedFlows will be totally removed, so getRunningFlows will only return runningFlows.

File: azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java [hunk 6]
```diff
@@ -497,10 +498,13 @@ private void loadQueuedFlows() throws ExecutorManagerException {
   @Override
   public List<Pair<ExecutableFlow, Executor>> getActiveFlowsWithExecutor()
```
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java]: Same comment as above.

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java (hunk 1 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: It's called in getExecutableFlows().

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java (hunk 2 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: Both web server and executor will update active_executing_flows.

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java (hunk 1 above)
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: Um.. My intellij doesn't allow me to point to the caller.

**Comment by kunkun-tang** [comment]
I had a look at your code again. Looks like you are forcing loading queues from db, rather than from cache, right? So the cache still exists. You might not say that you are removing cache in this pull request.

File: azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java [hunk 7]
```diff
@@ -839,6 +844,35 @@ public void testActiveReference() throws Exception {
     Assert.assertFalse(activeFlows2.containsKey(flow2.getExecutionId()));
   }
 
+  @Test
+  public void testFetchActiveFlowByExecId() throws Exception {
```
Comment (kunkun-tang) [inline azkaban-common/src/test/java/azkaban/executor/JdbcExecutorLoaderTest.java]: The test looks great.

File: azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java (hunk 1 above)
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: My intellij reports another method `fetchJobInfoAttempts`, which also looks like never used.

Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/executor/JdbcExecutorLoader.java]: You are right. fetchJobInfoAttempts is never used. Also I double checked fetchNumExecutableFlows. Actually fetchNumExecutableFlows(int projectId, String flowId) is used but fetchNumExecutableFlows() is not. I'll remove the unused ones.

**Pull request merged**